### DISCUSSION TITLE
Oqtane.org Performance & Vulnerability Test

### DISCUSSION THREAD

**Discussion by Sharafudeen** [oqtane/oqtane.framework#1971] [General] [answered]
Hi,

I just run **performance test** with [https://snyk.io/website-scanner](https://snyk.io/website-scanner) 
Here you can find **performance results** https://webpagetest.org/result/220129_AiDcHC_4c040fcff84b193d8fe6567db8298d99

@sbwalker Nearly 4.5 seconds for first loading.. why oqtane.org used javascripts 13% ? 

1. Any known performance vulnerabilities occurs while using Oqtane with Bootstrap 5?
2. How Oqtane can prevents high security risks?
3. Additional security implementations required in Oqtane?

Hope you have considered more on security risks with latest vulnerabilities such as Log4js, etc.,

**Performance test in gif file:**

![video](https://user-images.githubusercontent.com/42769344/151658605-11364daa-1329-4626-8c0e-8fcd4e8b0157.gif)

**Performance test screenshot:**

![Web capture_29-1-2022_145515_webpagetest org](https://user-images.githubusercontent.com/42769344/151658694-ab6ee18c-2d5d-470f-978a-98a44607fc9c.jpeg)

**Comment by sbwalker** [comment, marked as answer]
@Sharafudeen some interesting results - thank you for sharing. Please note that the theme for oqtane.org has not been optimized for performance at all at this point. For example, the header images are 1200X800px in size and have not been optimized for web.

- why oqtane.org used javascripts 13% ?

Blazor applications still use JavaScript. You will notice the largest JavasScript file is blazor.server.js - which provides the functionality for Blazor Server. Oqtane also has its own minimal set of JavaScript methods which it utilizes through JS Interop. And Bootstrap 5 also uses some JavaScript for popups, etc... so the theme includes the bootstrap.bundle.min.js. 

- Any known performance vulnerabilities occurs while using Oqtane with Bootstrap 5?

I have never heard of a "performance" vulnerability - did you mean "security" vulnerability? I have not heard of any security vulnerabilities based on Bootstrap 5.

- How Oqtane can prevents high security risks?

Just to be clear, ALL software is susceptible to security risks. One of the ways Oqtane mitigates this is to minimize third party dependencies - this is part of the Oqtane Philosophy - see https://www.oqtane.org/blog/!/20/oqtane-philosophy. The fewer dependencies, the fewer components in your tool chain which can be compromised. Blazor itself is also designed to be very secure.

- Additional security implementations required in Oqtane?

You should take all of the standard precautions when deploying Oqtane as you would with any web application. Many application vulnerabilities occur because the environment in which they are deployed does not follow proper security guidance.

- Hope you have considered more on security risks with latest vulnerabilities such as Log4js, etc.

The Log4j exploit highlighted the need for proper management of third party dependencies in software (which I mentioned above) and the fact that many people do not follow guidance to upgrade their dependencies on a regular basis - leaving them susceptible to attacks. A lot of effort has gone into streamlining the upgrade process in Oqtane so that you can keep your framework updated.

**Comment by sbwalker** [comment]
@Sharafudeen you said "I just run performance test with https://snyk.io/website-scanner" - but I do not see a performance test on their site - only a vulnerability scanner?

**Reply by Sharafudeen** [reply]
When I run vulnerability test using their website scanner and it is loading all oqtane.org site's assets. They have mentioned each requests & response time - loading speed / performance (refer test result link). Blazor always takes more time for FIRST LOADING? Any concerns on this in Oqtane?

**Reply by sbwalker** [reply]
Performance results depend on many factors including the environment, extensions, dependencies, etc... In the case of oqtane.org it is running on Blazor Server currently. If it was running on Blazor WebAssembly it would produce different results. Either way I personally have no concerns about the performance - especially on the administrative side, as the former DNN version of the site was painfully slow to manage.

**Reply by sbwalker** [reply]
I should also probably mention that it's not the first load scenario where you would notice the performance benefit of Blazor - it's on all user interactions AFTER the first load (as it does not need to do a full round trip to the server for every page request).
